### ISSUE TITLE
Can't run phone authentication tests against emulator because the sms code changes all the time

### ISSUE THREAD

**Issue by cedvdb** [firebase/firebase-tools#4290]
Currently the emulator returns a random String code when authenticating with phone against it. You can use this code to authenticate.

However for testing it would be much easier if the code was predictable, either via a configuration option or just a static code usage.

I've submitted a PR https://github.com/firebase/firebase-tools/pull/4255 that makes the code to always be '123456' but I do not know if that's the best option here.

**Comment by yuchenshi** [comment]
@cedvdb The verification code in the emulator can be retrieved via a simple REST call for non-interactive testing as documented here: https://firebase.google.com/docs/emulator-suite/connect_auth#non-interactive_testing_2

The reason behind random digits is mostly to be closer to production, especially when developers want to test 2+ concurrent sign-ins. If the two API calls return the same digits, then it is hard to test the logic to use  the correct verification code for the right session. Hope that makes sense.

**Comment by yuchenshi** [comment]
Would it unblock you if we support [something like this in production](https://firebase.google.com/docs/auth/web/phone-auth#test-with-fictional-phone-numbers), where you can predefine *some* phone numbers (of your choice) to have fixed verification codes? You can set them to 123456 if you like to, but you don't have to.

**Comment by cedvdb** [comment]
If the endpoint works I suppose this is resolved. I've implemented the retrieval but I have other tasks first before trying it out.

**Comment by cedvdb** [comment]
@yuchenshi I've tested the rest api to get the verification code. It works fine, although in practice I think it would be easier if it worked like the non emulator, where you can preset the sms code. Now I have two process depending if my tests run with emulator or the real firebase auth.

**Comment by cedvdb** [comment]
> @cedvdb The verification code in the emulator can be retrieved via a simple REST call for non-interactive testing as documented here: https://firebase.google.com/docs/emulator-suite/connect_auth#non-interactive_testing_2

This seems to work well for the first authentication, however when running automated tests, the second authentication does not regenerate a new code. I've searched a long time but cannot figure out why. I'll skip those tests for now.

**Comment by yuchenshi** [comment]
@cedvdb Is there any chance you can get a step-by-step repro? We'll gladly look into this.

**Comment by cedvdb** [comment]
I made a repro

https://github.com/cedvdb/flutter_repros/tree/failing_auth

However it might require some setup:

  - it uses flutter
  - it will require that you remove the firebase folder and initialize with your own project so you can run the emulator with `firebase init`
  - You'll have to change the values in `firebase_options.dart  android` with the values for android.


Then you can run the project on an emulator and you will see that the second time you login (press sign in, then sign out, then sign in again), the emulator does not regenerate a sms code.

**Comment by cedvdb** [comment]
@yuchenshi have you had the possibility of using the repro ?

**Comment by yuchenshi** [comment]
@cedvdb Thanks for sharing your findings about the issue not being present in web. That hints at issue being in some specific interactions between Flutter and Auth emulator, and we're waiting for my colleagues with Flutter expertise to take a look.

On the other hand though, I think your feature request is a perfectly reasonable one. As we strive to make the Auth emulator align with production behavior, we'd like to make static SMS code **opt-in** just like [production](https://firebase.google.com/docs/auth/web/phone-auth#test-with-fictional-phone-numbers). As such, we're open to reviewing your PR if you could modify it to allow such [configuration using the same API as production](https://cloud.google.com/identity-platform/docs/reference/rest/v2/projects/updateConfig).

While this may sounds like an extra step for your particular test setup, it actually helps us stay in sync with production, which unblocks many other developers from using the same setup script for both production and emulator-based integration testing.

LMK what you think! The config implementation can live somewhere here, BTW:

https://github.com/firebase/firebase-tools/blob/4ee276e0775fbb43d777ea4e1237f94f966d4a6a/src/emulator/auth/operations.ts#L2003

**Comment by yuchenshi** [comment]
@cedvdb Sorry to hear you're having troubles here. We take tackling these kind of roadblocks seriously and I'll be relaying your feedback to the right team too.

**Comment by yuchenshi** [comment]
@mikehardy Would you mind opening a separate issue with more details on generating a code the second time? A few steps or a repro will help a lot.
